nexus/src/app/sagas/image_create.rs

```diff
@@ -0,0 +1,618 @@
+                version: params.create_params.version.clone(),
+                digest: None, // TODO
+                block_size: source_volume.block_size,
+                size: source_volume.size.into(),
+            }
+        }
+
+        params::ImageSource::YouCanBootAnythingAsLongAsItsAlpine => {
+            db::model::Image {
+                identity: db::model::ImageIdentity::new(
+                    image_id,
+                    params.create_params.identity.clone(),
+                ),
+                silo_id: params.image_type.silo_id(),
+                project_id: params.image_type.project_id(),
+                volume_id: source_volume.volume_id.into(),
+                url: None,
+                os: "alpine".into(),
+                version: "propolis-blob".into(),
+                digest: None,
+                block_size: source_volume.block_size,
+                size: source_volume.size.into(),
+            }
+        }
+    };
+
+    match &params.image_type {
+        ImageType::Project { authz_project, .. } => osagactx
+            .datastore()
+            .project_image_create(
+                &opctx,
+                &authz_project,
+                record.try_into().map_err(ActionError::action_failed)?,
+            )
+            .await
+            .map_err(ActionError::action_failed),
+
+        ImageType::Silo { authz_silo, .. } => osagactx
+            .datastore()
+            .silo_image_create(
+                &opctx,
+                &authz_silo,
+                record.try_into().map_err(ActionError::action_failed)?,
+            )
+            .await
+            .map_err(ActionError::action_failed),
+    }
+}
+
+async fn simc_create_image_record_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    let image_id = sagactx.lookup::<Uuid>("image_id")?;
+
+    match &params.image_type {
+        ImageType::Project { .. } => {
+            let (.., authz_image, db_image) =
+                LookupPath::new(&opctx, &osagactx.datastore())
+                    .project_image_id(image_id)
+                    .fetch()
+                    .await?;
+
+            osagactx
+                .datastore()
+                .project_image_delete(&opctx, &authz_image, db_image)
+                .await?;
+        }
+
+        ImageType::Silo { .. } => {
+            let (.., authz_image, db_image) =
+                LookupPath::new(&opctx, &osagactx.datastore())
+                    .silo_image_id(image_id)
+                    .fetch()
+                    .await?;
+
+            osagactx
+                .datastore()
+                .silo_image_delete(&opctx, &authz_image, db_image)
+                .await
+                .map_err(ActionError::action_failed)?;
```

Comment:
nit: just use `?` here instead of mapping, to match the other error cases?

Reply:
thanks! 59345ff94